sql/tests: use of Span after Finish in the internal executor

sql/tests.TestRandomSyntaxFuncCommon [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/16180476?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/16180476?buildTab=artifacts#/) on release-24.2 @ [60d3fb84a73f5e31c99404223f3a0ba91e67f489](https://github.com/cockroachdb/cockroach/commits/60d3fb84a73f5e31c99404223f3a0ba91e67f489):


```
*   | 	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:182
*   | github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).RecordingType
*   | 	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:442
*   | github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).IsVerbose
*   | 	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:450
*   | github.com/cockroachdb/cockroach/pkg/util/log.ExpensiveLogEnabled
*   | 	github.com/cockroachdb/cockroach/pkg/util/log/log.go:64
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).sessionEventf
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:4339
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).close
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1247
*   | github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:238
*   | github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx.func2
*   | 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:264
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
*   | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:485
*   | runtime.goexit
*   | 	src/runtime/asm_amd64.s:1695
* Wraps: (4) use of Span after Finish. Span: flow. Finish previously called at: <stack not captured. Set debugUseAfterFinish>
* Error types: (1) *withstack.withStack (2) *assert.withAssertionFailure (3) *withstack.withStack (4) *errutil.leafError
*
panic: use of Span after Finish. Span: flow. Finish previously called at: <stack not captured. Set debugUseAfterFinish> [recovered]
	panic: use of Span after Finish. Span: flow. Finish previously called at: <stack not captured. Set debugUseAfterFinish>

goroutine 5881713 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x454825?, {0x827fa70, 0xc0274c9d10})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0x65
panic({0x62b3bc0?, 0xc00301c0b0?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish(0xc6790a0?)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:182 +0x114
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).RecordingType(0xc00c462c08)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:442 +0x18
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).IsVerbose(...)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:450
github.com/cockroachdb/cockroach/pkg/util/log.ExpensiveLogEnabled({0x827fa70?, 0xc0274c9380?}, 0x2)
	github.com/cockroachdb/cockroach/pkg/util/log/log.go:64 +0x34
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).sessionEventf(0xc0081d1578?, {0x827fa70, 0xc0274c9380}, {0x698aa49, 0x16}, {0x0, 0x0, 0x0})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:4339 +0x51
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).close(0xc007adaa08, {0x827fa70, 0xc0274c9380}, 0x2)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1247 +0x57
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx.func1()
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:238 +0x3b
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx.func2({0x827fa70, 0xc0274c9d10})
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:264 +0x124
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:485 +0x13a
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 5880972
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:476 +0x3fe
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #125113 sql/tests: TestRandomSyntaxFuncCommon failed [invalid datum type given: STRING COLLATE kms, expected STRING COLLATE kms] [C-test-failure O-robot P-3 T-sql-queries branch-master branch-release-22.2 branch-release-23.1 branch-release-23.1.23-rc]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxFuncCommon.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-40507

--- a/pkg/sql/internal.go
+++ b/pkg/sql/internal.go
@@ -230,7 +230,7 @@ func (ie *InternalExecutor) runWithEx(
 		return err
 	}
 	wg.Add(1)
-	cleanup := func() {
+	cleanup := func(ctx context.Context) {
 		closeMode := normalClose
 		if txn != nil {
 			closeMode = externalTxnClose
@@ -245,7 +245,7 @@ func (ie *InternalExecutor) runWithEx(
 			SpanOpt:  stop.ChildSpan,
 		},
 		func(ctx context.Context) {
-			defer cleanup()
+			defer cleanup(ctx)
 			// TODO(yuzefovich): benchmark whether we should be growing the
 			// stack size unconditionally.
 			if growStackSize {
@@ -265,7 +265,7 @@ func (ie *InternalExecutor) runWithEx(
 	); err != nil {
 		// The goroutine wasn't started, so we need to perform the cleanup
 		// ourselves.
-		cleanup()
+		cleanup(ctx)
 		return err
 	}
 	return nil

